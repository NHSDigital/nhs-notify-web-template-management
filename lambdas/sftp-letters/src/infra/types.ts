export type SftpSupplierConfig = {
  host: string;
  username: string;
  privateKey: string;
  hostKey: string;
  baseUploadDir: string;
  baseDownloadDir: string;
};
