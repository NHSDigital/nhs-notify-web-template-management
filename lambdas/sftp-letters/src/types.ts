import { Language, LetterType } from 'nhs-notify-backend-client';

export type ExpandedIdComponents = {
  clientId: string;
  templateId: string;
  campaignId: string;
  language: Language;
  letterType: LetterType;
};
