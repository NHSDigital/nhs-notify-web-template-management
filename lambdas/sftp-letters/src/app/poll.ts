import type { Logger } from 'nhs-notify-web-template-management-utils/logger';
import type { SftpSupplierClientRepository } from '../infra/sftp-supplier-client-repository';
import type { SftpClient } from '../infra/sftp-client';
import type { S3Repository } from 'nhs-notify-web-template-management-utils';

const MAX_SIZE_BYTES = 100 * 1024 * 1024; // 100MB in bytes;

export class App {
  constructor(
    private readonly sftpSupplierClientRepository: SftpSupplierClientRepository,
    private readonly logger: Logger,
    private readonly s3Repository: S3Repository,
    private readonly sftpEnvironment: string
  ) {}

  private async copyFolder(
    sftpClient: SftpClient,
    copyPath: string,
    pastePath: string
  ) {
    const isDir = (await sftpClient.exists(copyPath)) === 'd';

    if (!isDir) {
      this.logger.info(`Path '${copyPath}' does not exist`);
      return;
    }

    const sftpFiles = await sftpClient.list(copyPath);

    for (const sftpFile of sftpFiles) {
      if (sftpFile.type === 'd') {
        this.logger.info({
          description: 'Copying folder',
          copyPath,
          pastePath,
        });
        await this.copyFolder(
          sftpClient,
          `${copyPath}/${sftpFile.name}`,
          `${pastePath}/${sftpFile.name}`
        );
      }

      if (sftpFile.type === '-') {
        this.logger.info({ description: 'Copying file', copyPath, pastePath });
        await this.copyFile(
          sftpClient,
          `${copyPath}/${sftpFile.name}`,
          `${pastePath}/${sftpFile.name}`
        );
      }
    }
  }

  private async validateFile(buffer: Buffer) {
    return (
      buffer.byteLength > 0 &&
      buffer.byteLength < MAX_SIZE_BYTES &&
      this.isBufferPdf(buffer)
    );
  }

  isBufferPdf(buffer: Buffer): boolean {
    const pdfBufferStart = [0x25, 0x50, 0x44, 0x46]; // '%PDF'
    return pdfBufferStart.every((byte, index) => buffer[index] === byte);
  }

  private async copyFile(
    sftpClient: SftpClient,
    copyPath: string,
    pastePath: string
  ) {
    try {
      const data = (await sftpClient.get(copyPath)) as Buffer;

      const fileValidation = await this.validateFile(data);

      if (!fileValidation) {
        this.logger.error('PDF file failed validation', {
          copyPath,
          pastePath,
        });
        return;
      }

      await this.s3Repository.putRawData(data, pastePath);

      await sftpClient.delete(copyPath);
    } catch (error) {
      this.logger.error('Failed to copy file', { copyPath, pastePath, error });
    }
  }

  async poll() {
    const sftpConfigs = await this.sftpSupplierClientRepository.listClients();

    for (const { sftpClient, baseDownloadDir, name } of sftpConfigs) {
      this.logger.info('Polling SFTP');
      await sftpClient.connect();

      this.logger.info(
        `Copying files from folder ${baseDownloadDir} to proofs/${name}`
      );

      await this.copyFolder(
        sftpClient,
        `${baseDownloadDir}/${this.sftpEnvironment}/proofs`,
        'proofs'
      );

      await sftpClient.end();

      this.logger.info('Finished polling SFTP');
    }
  }
}
