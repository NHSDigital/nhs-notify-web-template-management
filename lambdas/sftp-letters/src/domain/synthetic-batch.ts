import { format } from 'date-fns';
import { createHash } from 'node:crypto';
import {
  pdsPersonalisationKeys,
  staticPdsExampleData,
} from './static-batch-data';

export type Manifest = {
  template: string;
  batch: string;
  records: string;
  md5sum: string;
};

export class SyntheticBatch {
  constructor(
    private readonly randomId: () => string,
    private readonly getDate: () => Date
  ) {}

  buildBatch(
    templateId: string,
    fields: string[],
    userTestData?: Record<string, string>[]
  ): Record<string, string>[] {
    const date = this.getDate();

    return Array.from({ length: 3 }, (_, i) => {
      const fieldEntries = [
        ['clientRef', this.clientRef(date)],
        ['template', templateId],
        ...fields.map((field) => {
          const value = this.fieldValue(
            field,
            date,
            staticPdsExampleData[i],
            userTestData?.[i]
          );
          return [field, value];
        }),
      ];

      return Object.fromEntries(fieldEntries);
    });
  }

  private fieldValue(
    field: string,
    date: Date,
    pdsData: Record<string, string>,
    userData?: Record<string, string>
  ) {
    if (field === 'date') {
      return format(date, 'd LLLL yyyy');
    }
    if (pdsPersonalisationKeys.includes(field)) {
      return pdsData[field];
    }
    return userData?.[field];
  }

  private clientRef(date: Date) {
    return [
      this.randomId(),
      this.randomId(),
      date.getTime().toString().slice(0, 10),
    ].join('_');
  }

  getId(templateId: string, pdfVersion: string) {
    const pseudoRandomSegment = pdfVersion.replaceAll('-', '').slice(0, 27);
    return `${templateId}-0000000000000_${pseudoRandomSegment}`;
  }

  getHeader(fields: string[]) {
    return `clientRef,template,${fields.join(',')}`;
  }

  buildManifest(
    templateId: string,
    batchId: string,
    batchCsv: string
  ): Manifest {
    return {
      template: templateId,
      batch: `${batchId}.csv`,
      records: '3',
      // eslint-disable-next-line sonarjs/hashing
      md5sum: createHash('md5').update(batchCsv).digest('hex'),
    };
  }
}
