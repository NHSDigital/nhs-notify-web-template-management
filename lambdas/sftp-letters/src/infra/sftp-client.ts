import type { Readable } from 'node:stream';
import Client, { type ConnectOptions } from 'ssh2-sftp-client';
import crypto from 'node:crypto';

type FileInfoType = 'd' | '-' | 'l';

export type FileInfo = {
  name: string;
  type: FileInfoType;
  modifyTime: number;
};

export class SftpClient {
  private readonly _client: Client;

  private readonly _config: ConnectOptions;

  constructor(
    host: string,
    username: string,
    privateKey: string,
    hostKey: string
  ) {
    this._client = new Client();
    this._config = {
      host,
      username,
      privateKey,
      hostVerifier: (key: Buffer): boolean =>
        hostKey ===
        `SHA256:${crypto.createHash('sha256').update(key).digest('base64')}`,
    };
  }

  public async connect(): Promise<void> {
    await this._client.connect(this._config);
  }

  public async end(): Promise<void> {
    await this._client.end();
  }

  public async put(objectStream: Readable, location: string): Promise<void> {
    await this._client.put(objectStream, location);
  }

  public async mkdir(path: string, recursive?: boolean): Promise<void> {
    await this._client.mkdir(path, recursive);
  }

  public async list(location: string): Promise<FileInfo[]> {
    const files = await this._client.list(location);
    return files.map(({ name, type, modifyTime }) => ({
      name,
      type,
      modifyTime,
    }));
  }

  public async exists(location: string): Promise<false | FileInfoType> {
    return this._client.exists(location);
  }

  public async get(
    location: string,
    destination?: string | NodeJS.WritableStream
  ): Promise<string | Buffer | NodeJS.WritableStream> {
    return await this._client.get(location, destination);
  }

  public async delete(location: string): Promise<void> {
    await this._client.delete(location);
  }

  public async rename(source: string, destination: string): Promise<void> {
    await this._client.rename(source, destination);
  }
}
