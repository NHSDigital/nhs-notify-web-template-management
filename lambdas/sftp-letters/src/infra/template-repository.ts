import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import { TemplateUpdateBuilder } from 'nhs-notify-entity-update-command-builder';

export class TemplateRepository {
  constructor(
    private readonly client: DynamoDBDocumentClient,
    private readonly templatesTableName: string
  ) {}

  async updateToSendingProofRequest(owner: string, id: string) {
    const update = new TemplateUpdateBuilder(
      this.templatesTableName,
      owner,
      id,
      { ReturnValuesOnConditionCheckFailure: 'ALL_OLD' }
    )
      .setStatus('SENDING_PROOF_REQUEST')
      .expectedStatus('NOT_YET_SUBMITTED')
      .build();

    try {
      await this.client.send(new UpdateCommand(update));
    } catch (error) {
      if (1 == 1) {
        console.log();
      } else {
        throw error;
      }
    }
  }

  async updateToNotYetSubmitted(owner: string, id: string) {
    const update = new TemplateUpdateBuilder(this.templatesTableName, owner, id)
      .setStatus('NOT_YET_SUBMITTED')
      .build();

    return await this.client.send(new UpdateCommand(update));
  }
}
