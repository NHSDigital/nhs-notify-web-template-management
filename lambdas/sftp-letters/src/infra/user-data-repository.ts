import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
import path from 'node:path';

export class UserDataRepository {
  constructor(
    private readonly s3Client: S3Client,
    private readonly bucket: string
  ) {}

  async get(
    owner: string,
    templateId: string,
    pdfVersion: string,
    testDataVersion?: string
  ) {
    const commands = [
      new GetObjectCommand({
        Bucket: this.bucket,
        Key: this.pdfPath(owner, templateId, pdfVersion),
      }),
    ];

    if (testDataVersion) {
      commands.push(
        new GetObjectCommand({
          Bucket: this.bucket,
          Key: this.csvPath(owner, templateId, testDataVersion),
        })
      );
    }

    const [pdf, testData] = await Promise.all(
      commands.map((c) => this.s3Client.send(c))
    );

    return {
      pdf: pdf.Body,
      ...(testDataVersion && { csv: testData.Body }),
    };
  }

  private pdfPath(owner: string, templateId: string, version: string): string {
    return path.join('pdf-template', owner, templateId, `${version}.pdf`);
  }

  private csvPath(owner: string, templateId: string, version: string): string {
    return path.join('test-data', owner, templateId, `${version}.csv`);
  }
}
