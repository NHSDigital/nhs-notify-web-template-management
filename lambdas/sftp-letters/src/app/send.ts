import type { SftpClient } from '../infra/sftp-client';
import type { UserDataRepository } from '../infra/user-data-repository';
import type { Logger } from 'nhs-notify-web-template-management-utils/logger';
import type { Batch } from '../domain/batch';
import type { TemplateRepository } from '../infra/template-repository';
import { parseTestPersonalisation } from '../domain/test-data';
import { serialiseCsv } from '../infra/serialise-csv';
import { z } from 'zod';
import path from 'node:path';
import { Readable } from 'node:stream';

function parseProofingRequest(event: string) {
  return z
    .object({
      owner: z.string(),
      templateId: z.string(),
      pdfVersion: z.string(),
      testDataVersion: z.string().optional(),
      fields: z.array(z.string()),
    })
    .parse(JSON.parse(event));
}

export class App {
  constructor(
    private readonly userDataRepository: UserDataRepository,
    private readonly templateRepository: TemplateRepository,
    private readonly sftpEnvironment: string,
    private readonly batch: Batch,
    private readonly logger: Logger
  ) {}
  async send(eventBody: string, sftpClient: SftpClient, baseUploadDir: string) {
    const { owner, templateId, pdfVersion, testDataVersion, fields } =
      parseProofingRequest(eventBody);

    const batchId = this.batch.getId(templateId, pdfVersion);

    const templateLogger = this.logger.child({
      owner,
      templateId,
      pdfVersion,
      testDataVersion,
      batchId,
    });

    templateLogger.info('Fetching user Data');

    const userData = await this.userDataRepository.get(
      owner,
      templateId,
      pdfVersion,
      testDataVersion
    );

    const parsedTestData = userData.testData
      ? parseTestPersonalisation(userData.testData)
      : undefined;

    const batchRows = this.batch.buildBatch(templateId, fields, parsedTestData);

    const batchHeader = this.batch.getHeader(fields);

    const batchCsv = await serialiseCsv(batchRows, batchHeader);

    const manifest = this.batch.buildManifest(templateId, batchId, batchCsv);

    const manifestCsv = await serialiseCsv(
      [manifest],
      'template,batch,records,md5sum'
    );

    const sftpEnvDir = path.join(baseUploadDir, this.sftpEnvironment);

    const templateDir = path.join(sftpEnvDir, 'templates', templateId);
    const batchDir = path.join(sftpEnvDir, 'batches', templateId);

    const pdfName = `${templateId}.pdf`;
    const csvName = `${batchId}.csv`;
    const manifestName = `${batchId}_MANIFEST.csv`;

    const pdfDestination = path.join(templateDir, pdfName);
    const batchDestination = path.join(batchDir, csvName);
    const manifestDestination = path.join(batchDir, manifestName);

    const batchStream = Readable.from(batchCsv);
    const manifestStream = Readable.from(manifestCsv);

    templateLogger.info('Updating template to SENDING_PROOF_REQUEST');

    if (await sftpClient.exists(manifestDestination)) {
      templateLogger.warn('Manifest already exists, assuming duplicate event');
      return;
    }

    templateLogger.info('Sending PDF');

    await Promise.all([
      sftpClient.mkdir(templateDir, true),
      sftpClient.mkdir(batchDir, true),
    ]);

    await sftpClient.put(userData.pdf, pdfDestination);

    templateLogger.info('Sending batch');

    await sftpClient.put(batchStream, batchDestination);

    templateLogger.info('Sending manifest');

    await sftpClient.put(manifestStream, manifestDestination);

    templateLogger.info('Updating template to NOT_YET_SUBMITTED');

    await this.templateRepository.updateToNotYetSubmitted(owner, templateId);

    templateLogger.info('Sent proofing request');
  }
}
