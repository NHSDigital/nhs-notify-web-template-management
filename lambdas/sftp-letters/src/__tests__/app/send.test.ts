import { mock } from 'jest-mock-extended';
import { TemplateRepository } from '../../infra/template-repository';
import { UserDataRepository } from '../../infra/user-data-repository';
import { App } from '../../app/send';
import { SyntheticBatch, Manifest } from '../../domain/synthetic-batch';
import { SftpClient } from '../../infra/sftp-client';
import { Readable } from 'node:stream';
import { mockTestDataCsv, streamToString } from '../helpers';
import { createMockLogger } from 'nhs-notify-web-template-management-test-helper-utils/mock-logger';
import { ProofingRequest } from '../../infra/types';

const sftpEnvironment = 'nhs-notify-web-template-management-main-app-api';
const baseUploadDir = 'Incoming';
const owner = 'owner-id';
const templateId = 'template-id';
const pdfVersion = 'pdf-version-id';
const testDataVersion = 'test-data-version-id';
const messageId = 'message-id';

function setup() {
  const userDataRepository = mock<UserDataRepository>();
  const templateRepository = mock<TemplateRepository>();
  const syntheticBatch = mock<SyntheticBatch>();
  const { logger, logMessages } = createMockLogger();

  const sftpClient = mock<SftpClient>();

  const app = new App(
    userDataRepository,
    templateRepository,
    sftpEnvironment,
    syntheticBatch,
    logger
  );

  return {
    app,
    mocks: {
      userDataRepository,
      templateRepository,
      syntheticBatch,
      logger,
      sftpClient,
    },
    logMessages,
  };
}

function mockEvent(
  hasTestData: boolean,
  personalisationParameters: string[]
): ProofingRequest {
  return {
    owner,
    templateId,
    pdfVersion,
    ...(hasTestData && { testDataVersion }),
    personalisationParameters,
  };
}

describe('App', () => {
  test('calls dependencies to send a proofing request', async () => {
    const { app, mocks } = setup();

    const personalisationParameters = ['pdsField', 'custom1', 'custom2'];
    const batchColumns = [
      'clientRef',
      'template',
      ...personalisationParameters,
    ];

    const event = mockEvent(true, personalisationParameters);

    const pdfContent = 'mock PDF content';
    const pdf = Readable.from(pdfContent);

    const batchId = 'template-id-0000000000000_pdfversionid';

    const testData = [
      { custom1: 'short1', custom2: 'short2' },
      { custom1: 'medium1', custom2: 'medium2' },
      { custom1: 'long1', custom2: 'long2' },
    ];

    const batchData = [
      {
        clientRef: 'random1_random2_1744184100',
        template: templateId,
        pdsField: 'pdsVal1',
        custom1: 'short1',
        custom2: 'short2',
      },
      {
        clientRef: 'random3_random4_1744184100',
        template: templateId,
        pdsField: 'pdsVal2',
        custom1: 'medium1',
        custom2: 'medium2',
      },
      {
        clientRef: 'random5_random6_1744184100',
        template: templateId,
        pdsField: 'pdsVal3',
        custom1: 'long1',
        custom2: 'long2',
      },
    ];

    const batchCsv: string = [
      batchColumns.join(','),
      batchData
        .map((x) =>
          [
            `"${x.clientRef}"`,
            `"${x.template}"`,
            `"${x.pdsField}"`,
            `"${x.custom1}"`,
            `"${x.custom2}"`,
          ].join(',')
        )
        .join('\n'),
    ].join('\n');

    const batchHash = 'hash-of-batch-csv';

    const manifestData: Manifest = {
      template: templateId,
      batch: `${batchId}.csv`,
      records: '3',
      md5sum: batchHash,
    };

    const manifestCsv = [
      'template,batch,records,md5sum',
      `"${templateId}","${batchId}.csv","3","${batchHash}"`,
    ].join('\n');

    const testDataCsv = mockTestDataCsv(['custom1', 'custom2'], testData);

    mocks.syntheticBatch.getId.mockReturnValueOnce(batchId);

    mocks.userDataRepository.get.mockResolvedValueOnce({
      testData: testDataCsv,
      pdf,
    });

    mocks.syntheticBatch.buildBatch.mockReturnValueOnce(batchData);

    mocks.syntheticBatch.getHeader.mockReturnValueOnce(batchColumns.join(','));

    mocks.syntheticBatch.buildManifest.mockReturnValueOnce(manifestData);

    mocks.templateRepository.acquireLock.mockResolvedValueOnce(true);

    // manifest doesn't already exist
    mocks.sftpClient.exists.mockResolvedValueOnce(false);

    const res = await app.send(
      JSON.stringify(event),
      messageId,
      mocks.sftpClient,
      baseUploadDir
    );

    expect(res).toBe('sent');

    expect(mocks.userDataRepository.get).toHaveBeenCalledTimes(1);
    expect(mocks.userDataRepository.get).toHaveBeenCalledWith(
      owner,
      templateId,
      pdfVersion,
      testDataVersion
    );

    expect(mocks.syntheticBatch.buildBatch).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.buildBatch).toHaveBeenCalledWith(
      templateId,
      personalisationParameters,
      testData
    );

    expect(mocks.syntheticBatch.getHeader).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.getHeader).toHaveBeenCalledWith(
      personalisationParameters
    );

    expect(mocks.syntheticBatch.buildManifest).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.buildManifest).toHaveBeenCalledWith(
      templateId,
      batchId,
      batchCsv
    );

    expect(mocks.templateRepository.acquireLock).toHaveBeenCalledTimes(1);
    expect(mocks.templateRepository.acquireLock).toHaveBeenCalledWith(
      owner,
      templateId
    );

    expect(mocks.sftpClient.exists).toHaveBeenCalledTimes(1);
    expect(mocks.sftpClient.exists).toHaveBeenCalledWith(
      `${baseUploadDir}/${sftpEnvironment}/batches/${templateId}/${batchId}_MANIFEST.csv`
    );

    expect(mocks.sftpClient.mkdir).toHaveBeenCalledTimes(2);
    expect(mocks.sftpClient.mkdir).toHaveBeenCalledWith(
      `${baseUploadDir}/${sftpEnvironment}/templates/${templateId}`,
      true
    );
    expect(mocks.sftpClient.mkdir).toHaveBeenCalledWith(
      `${baseUploadDir}/${sftpEnvironment}/batches/${templateId}`,
      true
    );

    expect(mocks.sftpClient.put).toHaveBeenCalledTimes(3);

    const [pdfPutCall, batchPutCall, manifestPutCall] =
      mocks.sftpClient.put.mock.calls;

    const [pdfArg, pdfDestinationArg] = pdfPutCall;

    expect(await streamToString(pdfArg)).toEqual(pdfContent);
    expect(pdfDestinationArg).toBe(
      `${baseUploadDir}/${sftpEnvironment}/templates/${templateId}/${templateId}.pdf`
    );

    const [batchArg, batchDestinationArg] = batchPutCall;

    expect(await streamToString(batchArg)).toEqual(batchCsv);
    expect(batchDestinationArg).toBe(
      `${baseUploadDir}/${sftpEnvironment}/batches/${templateId}/${batchId}.csv`
    );

    const [manifestArg, manifestDestinationArg] = manifestPutCall;

    expect(await streamToString(manifestArg)).toEqual(manifestCsv);
    expect(manifestDestinationArg).toBe(
      `${baseUploadDir}/${sftpEnvironment}/batches/${templateId}/${batchId}_MANIFEST.csv`
    );

    expect(mocks.templateRepository.clearLock).toHaveBeenCalledTimes(1);
    expect(mocks.templateRepository.clearLock).toHaveBeenCalledWith(
      owner,
      templateId
    );
  });

  test('throws if proof request event is invalid JSON', async () => {
    const { app, mocks } = setup();

    await expect(
      app.send('notjson', messageId, mocks.sftpClient, baseUploadDir)
    ).rejects.toThrow(`Unexpected token 'o', "notjson" is not valid JSON`);
  });

  test('throws if proof request event is missing a property', async () => {
    const { app, mocks } = setup();

    const { personalisationParameters: _, ...invalidEvent } = mockEvent(true, [
      'field',
    ]);

    await expect(
      app.send(
        JSON.stringify(invalidEvent),
        messageId,
        mocks.sftpClient,
        baseUploadDir
      )
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('exits early and does not send if a lock is already in effect', async () => {
    const { app, mocks, logMessages } = setup();

    const personalisationParameters = ['pdsField'];
    const batchColumns = [
      'clientRef',
      'template',
      ...personalisationParameters,
    ];

    const event = mockEvent(true, personalisationParameters);

    const pdfContent = 'mock PDF content';
    const pdf = Readable.from(pdfContent);

    const batchId = 'template-id-0000000000000_pdfversionid';

    const batchData = [
      {
        clientRef: 'random1_random2_1744184100',
        template: templateId,
        pdsField: 'pdsVal1',
      },
      {
        clientRef: 'random3_random4_1744184100',
        template: templateId,
        pdsField: 'pdsVal2',
      },
      {
        clientRef: 'random5_random6_1744184100',
        template: templateId,
        pdsField: 'pdsVal3',
      },
    ];

    const manifestData: Manifest = {
      template: templateId,
      batch: `${batchId}.csv`,
      records: '3',
      md5sum: 'hash-of-batch-csv',
    };

    mocks.syntheticBatch.getId.mockReturnValueOnce(batchId);
    mocks.userDataRepository.get.mockResolvedValueOnce({
      testData: undefined,
      pdf,
    });
    mocks.syntheticBatch.buildBatch.mockReturnValueOnce(batchData);
    mocks.syntheticBatch.getHeader.mockReturnValueOnce(batchColumns.join(','));
    mocks.syntheticBatch.buildManifest.mockReturnValueOnce(manifestData);

    // already locked
    mocks.templateRepository.acquireLock.mockResolvedValueOnce(false);

    const res = await app.send(
      JSON.stringify(event),
      messageId,
      mocks.sftpClient,
      baseUploadDir
    );

    expect(res).toBe('already-sent');

    expect(logMessages).toContainEqual(
      expect.objectContaining({
        level: 'warn',
        message: 'Template is already locked, assuming duplicate event',
      })
    );

    expect(mocks.userDataRepository.get).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.buildBatch).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.getHeader).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.buildManifest).toHaveBeenCalledTimes(1);

    expect(mocks.templateRepository.acquireLock).toHaveBeenCalledTimes(1);
    expect(mocks.templateRepository.acquireLock).toHaveBeenCalledWith(
      owner,
      templateId
    );

    expect(mocks.sftpClient.exists).not.toHaveBeenCalled();
    expect(mocks.sftpClient.mkdir).not.toHaveBeenCalled();
    expect(mocks.sftpClient.put).not.toHaveBeenCalled();
    expect(mocks.templateRepository.clearLock).not.toHaveBeenCalled();
  });

  test('exits early and does not send if the manifest is already in the SFTP server, clears existing lock', async () => {
    const { app, mocks, logMessages } = setup();

    const personalisationParameters = ['pdsField'];
    const batchColumns = [
      'clientRef',
      'template',
      ...personalisationParameters,
    ];

    const event = mockEvent(true, personalisationParameters);

    const pdfContent = 'mock PDF content';
    const pdf = Readable.from(pdfContent);

    const batchId = 'template-id-0000000000000_pdfversionid';

    const batchData = [
      {
        clientRef: 'random1_random2_1744184100',
        template: templateId,
        pdsField: 'pdsVal1',
      },
      {
        clientRef: 'random3_random4_1744184100',
        template: templateId,
        pdsField: 'pdsVal2',
      },
      {
        clientRef: 'random5_random6_1744184100',
        template: templateId,
        pdsField: 'pdsVal3',
      },
    ];

    const manifestData: Manifest = {
      template: templateId,
      batch: `${batchId}.csv`,
      records: '3',
      md5sum: 'hash-of-batch-csv',
    };

    mocks.syntheticBatch.getId.mockReturnValueOnce(batchId);
    mocks.userDataRepository.get.mockResolvedValueOnce({
      testData: undefined,
      pdf,
    });
    mocks.syntheticBatch.buildBatch.mockReturnValueOnce(batchData);
    mocks.syntheticBatch.getHeader.mockReturnValueOnce(batchColumns.join(','));
    mocks.syntheticBatch.buildManifest.mockReturnValueOnce(manifestData);

    // not already locked
    mocks.templateRepository.acquireLock.mockResolvedValueOnce(true);

    // but manifest already exists
    mocks.sftpClient.exists.mockResolvedValueOnce('-');

    const res = await app.send(
      JSON.stringify(event),
      messageId,
      mocks.sftpClient,
      baseUploadDir
    );

    expect(res).toBe('already-sent');

    expect(logMessages).toContainEqual(
      expect.objectContaining({
        level: 'warn',
        message: 'Manifest already exists, assuming duplicate event',
      })
    );

    expect(mocks.userDataRepository.get).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.buildBatch).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.getHeader).toHaveBeenCalledTimes(1);
    expect(mocks.syntheticBatch.buildManifest).toHaveBeenCalledTimes(1);
    expect(mocks.templateRepository.acquireLock).toHaveBeenCalledTimes(1);

    expect(mocks.sftpClient.exists).toHaveBeenCalledTimes(1);
    expect(mocks.sftpClient.exists).toHaveBeenCalledWith(
      `${baseUploadDir}/${sftpEnvironment}/batches/${templateId}/${batchId}_MANIFEST.csv`
    );

    expect(mocks.templateRepository.clearLock).toHaveBeenCalledTimes(1);
    expect(mocks.templateRepository.clearLock).toHaveBeenCalledWith(
      owner,
      templateId
    );

    expect(mocks.sftpClient.mkdir).not.toHaveBeenCalled();
    expect(mocks.sftpClient.put).not.toHaveBeenCalled();
  });

  test('logs handled errors', async () => {
    const { app, mocks, logMessages } = setup();

    const personalisationParameters = ['pdsField', 'custom1', 'custom2'];

    const event = mockEvent(true, personalisationParameters);

    const batchId = 'template-id-0000000000000_pdfversionid';

    mocks.syntheticBatch.getId.mockReturnValueOnce(batchId);

    const error = new Error('no PDF');

    mocks.userDataRepository.get.mockRejectedValueOnce(error);

    const res = await app.send(
      JSON.stringify(event),
      messageId,
      mocks.sftpClient,
      baseUploadDir
    );

    expect(res).toBe('failed');

    expect(mocks.userDataRepository.get).toHaveBeenCalledTimes(1);
    expect(mocks.userDataRepository.get).toHaveBeenCalledWith(
      owner,
      templateId,
      pdfVersion,
      testDataVersion
    );

    expect(logMessages).toContainEqual(
      expect.objectContaining({
        description: 'Failed to handle proofing request',
        level: 'error',
        batchId,
        message: error.message,
        messageId,
        owner,
        pdfVersion,
      })
    );
  });
});
