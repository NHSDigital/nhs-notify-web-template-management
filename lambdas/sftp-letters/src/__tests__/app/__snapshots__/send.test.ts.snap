// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App throws if proof request event is missing a property 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "personalisationParameters"
    ],
    "message": "Required"
  }
]"
`;
